x-pack/plugins/lens/public/react_embeddable/expressions/on_render.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { KibanaExecutionContext } from '@kbn/core-execution-context-common';
+import { canTrackContentfulRender } from '@kbn/presentation-containers';
+import { reportPerformanceMetricEvent } from '@kbn/ebt-tools';
+import { TableInspectorAdapter } from '../../editor_frame_service/types';
+
+import { getExecutionContextEvents, trackUiCounterEvents } from '../../lens_ui_telemetry';
+import { GetStateType, LensApi, LensEmbeddableStartServices, LensInternalApi } from '../types';
+import { getSuccessfulRequestTimings } from '../../report_performance_metric_util';
+import { addLog } from '../logger';
+
+function trackContentfulRender(activeData: TableInspectorAdapter, parentApi: unknown) {
+  if (!activeData || !canTrackContentfulRender(parentApi)) {
+    return;
+  }
+
+  const hasData = Object.values(activeData).some((table) => {
+    if (table.meta?.statistics?.totalCount != null) {
+      // if totalCount is set, refer to total count
+      return table.meta.statistics.totalCount > 0;
+    }
+    // if not, fall back to check the rows of the table
+    return table.rows.length > 0;
+  });
+
+  if (hasData) {
+    parentApi.trackContentfulRender();
+  }
+}
+
+function trackPerformanceMetrics(
+  api: LensApi,
+  coreStart: LensEmbeddableStartServices['coreStart']
+) {
+  const inspectorAdapters = api.getInspectorAdapters();
+  const timings = getSuccessfulRequestTimings(inspectorAdapters);
+  if (timings) {
+    const esRequestMetrics = {
+      eventName: 'lens_chart_es_request_totals',
+      duration: timings.requestTimeTotal,
+      key1: 'es_took_total',
+      value1: timings.esTookTotal,
+    };
+    reportPerformanceMetricEvent(coreStart.analytics, esRequestMetrics);
+  }
+}
+
+export function prepareOnRender(
+  api: LensApi,
+  internalApi: LensInternalApi,
+  parentApi: unknown,
+  getState: GetStateType,
+  { datasourceMap, visualizationMap, coreStart }: LensEmbeddableStartServices,
+  executionContext: KibanaExecutionContext | undefined,
+  dispatchRenderComplete: () => void
+) {
+  return function onRender$(count: number) {
```

Comment:
nit: do we need a named function here?

Reply:
It has been useful to debug it, particularly to see it in the stacktrace.

Reply:
all good so